File: server/backend/app/feedalyze/schema/topic.py
from datetime import datetime

from pydantic import Field

from backend.common.schema import SchemaBase


class TopicBase(SchemaBase):
    """需求主题基础模型"""

    title: str = Field(description='主题标题')
    category: str = Field(
        default='other',
        description='分类: bug=缺陷, improvement=体验优化, feature=新功能, performance=性能, other=其他'
    )
    status: str = Field(
        default='pending',
        description='状态: pending=待处理, planned=计划中, in_progress=进行中, completed=已完成, ignored=已忽略'
    )
    description: str | None = Field(None, description='详细描述')


class TopicCreate(TopicBase):
    """创建主题参数"""
    pass


class TopicUpdate(SchemaBase):
    """更新主题参数"""

    title: str | None = Field(None, description='主题标题')
    category: str | None = Field(None, description='分类')
    status: str | None = Field(None, description='状态')
    description: str | None = Field(None, description='详细描述')


class TopicOut(TopicBase):
    """主题输出模型"""

    id: str = Field(description='主题ID')
    tenant_id: str = Field(description='租户ID')
    ai_generated: bool = Field(description='是否AI生成')
    ai_confidence: float | None = Field(None, description='AI置信度 (0-1)')
    feedback_count: int = Field(description='关联反馈数量')
    created_time: datetime = Field(description='创建时间')
    updated_time: datetime = Field(description='更新时间')
    deleted_at: datetime | None = Field(None, description='删除时间')


class TopicDetailOut(TopicOut):
    """主题详情输出（包含关联数据）"""

    feedbacks: list['FeedbackOut'] = Field(default_factory=list, description='关联反馈列表')
    priority_score: 'PriorityScoreOut | None' = Field(None, description='优先级评分')
    status_history: list['StatusHistoryOut'] = Field(default_factory=list, description='状态变更历史')


class TopicListParams(SchemaBase):
    """主题列表查询参数"""

    status: str | None = Field(None, description='过滤：状态')
    category: str | None = Field(None, description='过滤：分类')
    sort_by: str = Field(default='created_time', description='排序字段: created_time, priority, feedback_count')
    sort_order: str = Field(default='desc', description='排序方向: asc, desc')


class TopicStatusUpdateParam(SchemaBase):
    """更新主题状态参数"""

    status: str = Field(description='新状态')
    reason: str | None = Field(None, description='变更原因')


# 避免循环导入，在文件末尾导入
from backend.app.feedalyze.schema.feedback import FeedbackOut  # noqa: E402
from backend.app.feedalyze.schema.priority import PriorityScoreOut  # noqa: E402
from backend.app.feedalyze.schema.status_history import StatusHistoryOut  # noqa: E402

TopicDetailOut.model_rebuild()

File: front/apps/web-antd/src/api/userecho/topic.ts
/**
 * 需求主题 API
 */

import type { PriorityScore } from './priority';

import { requestClient } from '#/api/request';

/** 主题对象 */
export interface Topic {
  id: string;
  tenant_id: string;
  board_id?: string;
  title: string;
  category: string;
  status: string;
  description?: string;
  ai_generated: boolean;
  ai_confidence?: number;
  feedback_count: number;
  centroid?: number[] | null;
  cluster_quality?: Record<string, any> | null;
  is_noise?: boolean;
  priority_score?: {
    id: string;
    impact_scope: number;
    business_value: number;
    dev_cost: number;
    urgency_factor: number;
    total_score: number;
    details?: any; // JSON
    tenant_id: string;
    topic_id: string;
    created_time: string;
    updated_time?: string;
  } | null;
  created_time: string;
  updated_time: string;
  deleted_at?: string;
}

/** 主题详情（包含关联数据） */
export interface TopicDetail {
  topic: Topic;
  feedbacks: Array<{
    id: string;
    content: string;
    customer_name?: string;
    submitted_at: string;
  }>;
  priority_score?: PriorityScore | null;
  status_history: Array<{
    id: string;
    from_status: string;
    to_status: string;
    reason?: string;
    changed_by: number;
    changed_by_name?: string;
    changed_at: string;
  }>;
}

/** 创建主题参数 */
export interface CreateTopicParams {
  title: string;
  category?: string;
  status?: string;
  description?: string;
  board_id?: string;
}

/** 更新主题参数 */
export interface UpdateTopicParams {
  title?: string;
  category?: string;
  status?: string;
  description?: string;
  board_id?: string;
}

/** 主题列表查询参数 */
export interface TopicListParams {
  skip?: number;
  limit?: number;
  status?: string[];
  category?: string[];
  board_ids?: string[];
  sort_by?: string;
  sort_order?: string;
  search_query?: string;
  search_mode?: 'keyword' | 'semantic';
}

/** 更新主题状态参数 */
export interface UpdateTopicStatusParams {
  status: string;
  reason?: string;
}

/**
 * 获取主题列表
 */
export async function getTopicList(params: TopicListParams) {
  return requestClient.get<Topic[]>('/api/v1/app/topics', { params });
}

/**
 * 获取主题详情
 */
export async function getTopicDetail(id: string) {
  return requestClient.get<TopicDetail>(`/api/v1/app/topics/${id}`);
}

/**
 * 创建主题
 */
export async function createTopic(data: CreateTopicParams) {
  return requestClient.post<Topic>('/api/v1/app/topics', data);
}

/**
 * 更新主题
 */
export async function updateTopic(id: string, data: UpdateTopicParams) {
  return requestClient.put<Topic>(`/api/v1/app/topics/${id}`, data);
}

/**
 * 更新主题状态
 */
export async function updateTopicStatus(id: string, data: UpdateTopicStatusParams) {
  return requestClient.put<Topic>(`/api/v1/app/topics/${id}/status`, data);
}

/**
 * 获取待确认主题数量
 */
export async function getPendingTopicCount() {
  return requestClient.get<{ count: number }>('/api/v1/app/topics/stats/pending-count');
}

/**
 * 删除主题
 */
export async function deleteTopic(id: string) {
  return requestClient.delete(`/api/v1/app/topics/${id}`);
}

/**
 * 关联反馈到主题
 */
export async function linkFeedbacksToTopic(topicId: string, feedbackIds: string[]) {
  return requestClient.post<{ count: number }>(`/api/v1/app/topics/${topicId}/feedbacks`, {
    feedback_ids: feedbackIds,
  });
}

/**
 * 从主题移除反馈关联
 */
export async function unlinkFeedbackFromTopic(topicId: string, feedbackId: string) {
  return requestClient.delete(`/api/v1/app/topics/${topicId}/feedbacks/${feedbackId}`);
}

/** 主题分类选项 */
export const TOPIC_CATEGORIES = [
  { value: 'bug', label: 'Bug', color: 'red' },
  { value: 'improvement', label: '体验优化', color: 'blue' },
  { value: 'feature', label: '新功能', color: 'green' },
  { value: 'performance', label: '性能问题', color: 'orange' },
  { value: 'other', label: '其他', color: 'default' },
];

/** 主题状态选项 */
export const TOPIC_STATUSES = [
  { value: 'pending', label: '待处理', color: 'default' },
  { value: 'planned', label: '已规划', color: 'blue' },
  { value: 'in_progress', label: '进行中', color: 'orange' },
  { value: 'completed', label: '已完成', color: 'green' },
  { value: 'ignored', label: '已忽略', color: 'red' },
];
